crypto/ml_dsa/ml_dsa_key_compress.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "ml_dsa_local.h"
+
+/* Key Compression related functions (Rounding & hints) */
+
+/**
+ * @brief Decompose r into (r1, r0) such that r == r1 * 2^13 + r0 mod q
+ * See FIPS 204, Algorithm 35, Power2Round()
+ *
+ * Note: that this code is more complex than the FIPS 204 spec since it keeps
+ * r0 as a positive number
+ *
+ * r mod +- 2^13 is defined as having a range of -4095..4096
+ *
+ * i.e for r = 0..4096 r1 = 0 and r0 = 0..4096
+ * at r = 4097..8191 r1 = 1 and r0 = -4095..0
+ * (but since r0 is kept positive it effectively adds q and then reduces by q if needed)
+ * Similarly for the range r = 8192..8192+4096 r1=1 and r0=0..4096
+ * & 12289..16383 r1=2 and r0=-4095..0
+ *
+ * @param r is in the range 0..q-1
+ * @param r1 The returned top 10 MSB (i.e it ranges from 0..1023)
+ * @param r0 The remainder in the range (0..4096 or q-4095..q-1)
+ *           So r0 has an effective range of 8192 (i.e. 13 bits).
+ */
+void ossl_ml_dsa_key_compress_power2_round(uint32_t r, uint32_t *r1, uint32_t *r0)
+{
+    unsigned int mask;
+    uint32_t r0_adjusted, r1_adjusted;
+
+    *r1 = r >> ML_DSA_D_BITS;         /* top 13 bits */
+    *r0 = r - (*r1 << ML_DSA_D_BITS); /* The remainder mod q */
+
+    r0_adjusted = mod_sub(*r0, 1 << ML_DSA_D_BITS);
+    r1_adjusted = *r1 + 1;
+
+    /* Mask is set iff r0 > (2^(dropped_bits))/2. */
+    mask = constant_time_lt((uint32_t)(1 << (ML_DSA_D_BITS - 1)), *r0);
+    /* r0 = mask ? r0_adjusted : r0 */
+    *r0 = constant_time_select_int(mask, r0_adjusted, *r0);
+    /* r1 = mask ? r1_adjusted : r1 */
+    *r1 = constant_time_select_int(mask, r1_adjusted, *r1);
+}
+
+/*
+ * @brief return the r1 component of Decomposing r into (r1, r0) such that
+ * r == r1 * (2 & gamma) + r0 mod q
+ * See FIPS 204, Algorithm 37, HighBits()
+ *
+ * @param r A value to decompose in the range (0..q-1)
+ * @param gamma2 Depending on the algorithm gamma2 is either (q-1)/32 or (q-1)/88
+ * @returns r1
+ */
+uint32_t ossl_ml_dsa_key_compress_high_bits(uint32_t r, uint32_t gamma2)
+{
+    uint32_t r1 = (r + 127) >> 7;
```

Comment:
```suggestion
    int32_t r1 = (r + 127) >> 7;
```
This only works correctly with *signed* integers.